go/arrow/compute/internal/kernels/vector_run_end_encode.go

```diff
@@ -0,0 +1,953 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build go1.18
+
+package kernels
+
+import (
+	"bytes"
+	"fmt"
+	"sort"
+	"unsafe"
+
+	"github.com/apache/arrow/go/v12/arrow"
+	"github.com/apache/arrow/go/v12/arrow/bitutil"
+	"github.com/apache/arrow/go/v12/arrow/compute/internal/exec"
+	"github.com/apache/arrow/go/v12/arrow/decimal128"
+	"github.com/apache/arrow/go/v12/arrow/decimal256"
+	"github.com/apache/arrow/go/v12/arrow/float16"
+	"github.com/apache/arrow/go/v12/arrow/internal/debug"
+	"github.com/apache/arrow/go/v12/arrow/memory"
+)
+
+type RunEndEncodeState struct {
+	RunEndType arrow.DataType
+}
+
+func (RunEndEncodeState) TypeName() string {
+	return "RunEndEncodeOptions"
+}
+
+type RunEndsType interface {
+	int16 | int32 | int64
+}
+
+func readFixedWidthVal[V exec.FixedWidthTypes](inputValidity, inputValues []byte, offset int64, out *V) bool {
+	if bitutil.BitIsNotSet(inputValidity, int(offset)) {
+		return false
+	}
+
+	sz := int64(unsafe.Sizeof(*out))
+	*out = *(*V)(unsafe.Pointer(&inputValues[offset*sz]))
```

Comment:
This could always be done without waiting for the load in `BitIsNotSet`, right? Less data dependency.

Reply:
fair point, I'm not sure if the Go compiler will optimize for this particular case, but it's worthwhile to help the CPU pipeline. I'll reorder this